src/MediatR/Mediator.cs

```diff
@@ -73,20 +88,65 @@ public void Publish(INotification notification)
 
         public Task PublishAsync(IAsyncNotification notification)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, notificationHandlers);
+            }
+        }
+
+        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        {
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, cancellationToken, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, cancellationToken, notificationHandlers);
+            }
+        }
+
+        private async Task PublishSequentialAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            foreach (var handler in notificationHandlers)
+            {
+                await handler.Handle(notification);
+            }
+        }
+
+        private Task PublishParallelAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification))
                 .ToArray();
 
-            return Task.WhenAll(notificationHandlers);
+            return Task.WhenAll(tasks);
         }
 
-        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        private Task PublishParallelAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken,
+            IEnumerable<CancellableAsyncNotificationHandlerWrapper> notificationHandlers)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification, cancellationToken))
                 .ToArray();

```

Comment:
> False or true, anyway let's be more friendly, right?

I'm sorry. It wasn't my intention to be unfriendly :confused: I just wanted to point out that what you said is incorrect and based on false assumptions. @samueldjack did nothing wrong by calling `ToArray` before passing the tasks to `WhenAll`, or simply leaving the code as-is. The sooner the tasks are  "kicked off", the better, IMO :smile:

> I'm basing my statement on my experience that calling ToArray() kick-offs these tasks before they're reaching Task.WhenAll() what doesn't help much during debugging.

Yes, the enumerable is enumerated when calling ´ToArray´, but does it matter if that happens before the call to `WhenAll`? It's only a matter of time before `ToArray` [is called anyway](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,6016)... If anything, you could argue that passing an array is marginally *faster* than calling the enumerable overload, because of fewer checks. I'm not sure how this affects debugging?